File: typescript/3.ts
function lengthOfLongestSubstring(s: string): number {
    let ret = 0;

    let front = 0;
    const cnt: Map<string, number> = new Map();
    for (let i = 0; i < s.length; i++) {
        const ch = s[i];
        if (!cnt.has(ch)) {
            ret = Math.max(ret, i - front + 1);
        } else {
            while (cnt.has(ch)) {
                cnt.delete(s[front]);
                front++;
            }
        }
        cnt.set(ch, 1);
    }

    return ret;
}

test('3', () => {
    expect(lengthOfLongestSubstring('bbbbbb')).toBe(1);
    expect(lengthOfLongestSubstring('tmmzuxt')).toBe(5);
});

File: python/3.py
import unittest
from typing import Dict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ret = 0

        front = 0
        cnt: Dict[str, int] = dict()
        for i in range(len(s)):
            ch = s[i]
            if ch not in cnt:
                ret = max(ret, i - front + 1)
            else:
                while ch in cnt:
                    del cnt[s[front]]
                    front += 1
            cnt[ch] = 1
        return ret


class TestSolution(unittest.TestCase):
    def test(self):
        solution = Solution()
        self.assertEqual(solution.lengthOfLongestSubstring("abcabcbb"), 3)
        self.assertEqual(solution.lengthOfLongestSubstring("tmmzuxt"), 5)
